### PR TITLE
0301 vm linux

### PR DESCRIPTION
**Pull request by mumian**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

**Comment by mumian** [comment]
@bmoore-msft - Hi Brian, sorry to give you another broken PR.  The error message doesn't help too much.  I have tested the template by using PowerShell from my local computer.  In the template, I "grouped" the variables.  I think it increases readability.  But I am not sure it is standard practice.  Thanks.

File: 101-vm-sshkey/azuredeploy.json [hunk 1]
```diff
@@ -2,47 +2,62 @@
...
+    "location": {
+      "type": "string",
+      "defaultValue": "[resourceGroup().location]",
+      "metadata": {
+        "description": "Specifies the location for all resources."
+      }
+    },
     "adminUsername": {
       "type": "string",
       "defaultValue": "azureuser",
```
Comment (bmoore-msft) [inline 101-vm-sshkey/azuredeploy.json]: remove the defaultValue

File: 101-vm-sshkey/README.md [hunk 2]
```diff
@@ -7,8 +7,26 @@
...
+This template allows you to create a Linux Virtual Machine with SSH Keys. This template also deploys Virtual Network with an inbound rule allowing only port 22 connections, Public IP addresses, a Public domain namespace, and a Network Interface. 
 
-NOTE: The SSH rsa public key is typically generated by ssh-keygen, among other tools, on Linux, Mac, or Windows. If you already have an ~/.ssh/id_rsa.pub file, that will work directly, as you would expect. When deployed using the Preview Portal, you should paste in the contents of your ssh-rsa public key file as a string. 
+The SSH rsa public key is typically generated by ssh-keygen, among other tools, on Linux, Mac, or Windows. If you already have an ~/.ssh/id_rsa.pub file, that will work directly, as you would expect. When deployed using the Preview Portal, you should paste in the contents of your ssh-rsa public key file as a string.
```
Comment (bmoore-msft) [inline 101-vm-sshkey/README.md]: what's the "preview portal"?

File: 101-vm-sshkey/README.md [hunk 3]
```diff
@@ -7,8 +7,26 @@
...
 
 When you deploy the template from the Azure CLI, when the parameter prompts you for the string, do not surround it in quotes, or they will be treated as part of the public key. 
+
+For more information about SSH key, see:
+
+- [Quick steps: Create and use an SSH public-private key pair for Linux VMs in Azure](https://docs.microsoft.com/azure/virtual-machines/linux/mac-create-ssh-keys)
+- [How to use SSH keys with Windows on Azure](https://docs.microsoft.com/azure/virtual-machines/linux/ssh-from-windows)
+- [Detailed steps: Create and manage SSH keys for authentication to a Linux VM in Azure](https://docs.microsoft.com/azure/virtual-machines/linux/create-ssh-keys-detailed)
+
+If you are new to Azure virtual machine, see:
```
Comment (bmoore-msft) [inline 101-vm-sshkey/README.md]: machines (s/b plural)

File: 101-vm-sshkey/azuredeploy.json [hunk 4]
```diff
@@ -2,47 +2,62 @@
...
-    "nicName": "sshNIC",
-    "publicIPAddressName": "sshPublicIP",
-    "publicIPAddressType": "Dynamic",
-    "virtualNetworkName": "sshVNET",
-    "networkSecurityGroupName": "networkSecurityGroup1",
-    "sshKeyPath": "[concat('/home/',parameters('adminUsername'),'/.ssh/authorized_keys')]",
-    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks', variables('virtualNetworkName'))]",
-    "subnet1Ref": "[concat(variables('vnetID'),'/subnets/',variables('subnet1Name'))]",
-    "apiVersion": "2015-06-15"
+    "networkInterface": {
```
Comment (bmoore-msft) [inline 101-vm-sshkey/azuredeploy.json]: let's leave the variables flat - the object variables add some complexity and tooling does not handle these well today.

There are a number of single use variables which we can just remove and replace with literals (like you did for vm image info)

File: 101-vm-sshkey/azuredeploy.json [hunk 5]
```diff
@@ -63,52 +78,47 @@
...
       "type": "Microsoft.Network/networkInterfaces",
-      "name": "[variables('nicName')]",
-      "location": "[variables('location')]",
+      "name": "[variables('networkInterface').name]",
+      "location": "[parameters('location')]",
+      "apiVersion": "2018-11-01",
       "dependsOn": [
-        "[concat('Microsoft.Network/publicIPAddresses/', variables('publicIPAddressName'))]",
-        "[concat('Microsoft.Network/virtualNetworks/', variables('virtualNetworkName'))]"
+        "[concat('Microsoft.Network/publicIPAddresses/', variables('publicIPAddress').name)]",
```
Comment (bmoore-msft) [inline 101-vm-sshkey/azuredeploy.json]: use resourceID instead of concat

File: 101-vm-sshkey/azuredeploy.json [hunk 6]
```diff
@@ -117,68 +127,68 @@
             "properties": {
               "privateIPAllocationMethod": "Dynamic",
               "publicIPAddress": {
-                "id": "[resourceId('Microsoft.Network/publicIPAddresses', variables('publicIPAddressName'))]"
+                "id": "[resourceId('Microsoft.Network/publicIPAddresses', variables('publicIPAddress').name)]"
               },
               "subnet": {
-                "id": "[variables('subnet1Ref')]"
+                "id": "[concat(resourceId('Microsoft.Network/virtualNetworks', variables('virtualNetwork').name),'/subnets/', variables('virtualNetwork').subnets.name)]"
```
Comment (bmoore-msft) [inline 101-vm-sshkey/azuredeploy.json]: remove concat and just use resourceId

File: 101-vm-sshkey/azuredeploy.json [hunk 7]
```diff
@@ -117,68 +127,68 @@
...
-      "apiVersion": "2016-04-30-preview",
       "type": "Microsoft.Compute/virtualMachines",
-      "name": "[variables('vmName')]",
-      "location": "[variables('location')]",
+      "name": "[variables('virtualMachine').name]",
+      "location": "[parameters('location')]",
+      "apiVersion": "2018-10-01",
       "dependsOn": [
-        "[concat('Microsoft.Network/networkInterfaces/', variables('nicName'))]"
+        "[concat('Microsoft.Network/networkInterfaces/', variables('networkInterface').name)]"
```
Comment (bmoore-msft) [inline 101-vm-sshkey/azuredeploy.json]: use resourceID

File: 101-vm-sshkey/azuredeploy.json (hunk 1 above)
Comment (mumian) [reply 101-vm-sshkey/azuredeploy.json]: fixed

File: 101-vm-sshkey/README.md (hunk 2 above)
Comment (mumian) [reply 101-vm-sshkey/README.md]: fixed

File: 101-vm-sshkey/README.md (hunk 3 above)
Comment (mumian) [reply 101-vm-sshkey/README.md]: fixed

File: 101-vm-sshkey/azuredeploy.json (hunk 4 above)
Comment (mumian) [reply 101-vm-sshkey/azuredeploy.json]: fixed. I used literals for the publicIPAddresses sku. Let me know it needs to be a parameter.

File: 101-vm-sshkey/azuredeploy.json (hunk 5 above)
Comment (mumian) [reply 101-vm-sshkey/azuredeploy.json]: fixed.

File: 101-vm-sshkey/azuredeploy.json (hunk 6 above)
Comment (mumian) [reply 101-vm-sshkey/azuredeploy.json]: fixed.

File: 101-vm-sshkey/azuredeploy.json (hunk 7 above)
Comment (mumian) [reply 101-vm-sshkey/azuredeploy.json]: fixed.

File: 101-vm-sshkey/azuredeploy.json [hunk 8]
```diff
@@ -2,47 +2,48 @@
...
-    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks', variables('virtualNetworkName'))]",
-    "subnet1Ref": "[concat(variables('vnetID'),'/subnets/',variables('subnet1Name'))]",
-    "apiVersion": "2015-06-15"
+    "vNetName": "[concat(parameters('projectName'), '-vnet')]",
+    "vNetAddressPrefixes": "10.0.0.0/16",
+    "vNetSubnetName": "default",
+    "vNetSubnetAddressPrefix": "10.0.0.0/24",
+    "vmName": "[concat(parameters('projectName'), '-vm')]",
+    "publicIPAddressName": "[concat(parameters('projectName'), '-ip')]",
+    "networkInterfaceName": "[concat(parameters('projectName'), subString(uniqueString(subscription().subscriptionId), 0, 3))]",
```
Comment (bmoore-msft) [inline 101-vm-sshkey/azuredeploy.json]: why does this need uniqueString()?

**Pull request merged**